packages/lodestar/src/db/attestationPool.ts

```diff
@@ -0,0 +1,230 @@
+ * will be refused.
+ */
+const SLOTS_RETAINED = 3;
+
+/**
+ * The maximum number of distinct `AttestationData` that will be stored in each slot.
+ *
+ * This is a DoS protection measure.
+ */
+const MAX_ATTESTATIONS_PER_SLOT = 16_384;
+
+type AggregateFast = {
+  data: phase0.Attestation["data"];
+  aggregationBits: boolean[];
+  signature: Signature;
+};
+
+/** Hex string of DataRoot `TODO` */
+type DataRootHex = string;
+
+/**
+ * A pool of `Attestation` that is specially designed to store "unaggregated" attestations from
+ * the native aggregation scheme.
+ *
+ * **The `NaiveAggregationPool` does not do any signature or attestation verification. It assumes
+ * that all `Attestation` objects provided are valid.**
+ *
+ * ## Details
+ *
+ * The pool sorts the `Attestation` by `attestation.data.slot`, then by `attestation.data`.
+ *
+ * As each unaggregated attestation is added it is aggregated with any existing `attestation` with
+ * the same `AttestationData`. Considering that the pool only accepts attestations with a single
+ * signature, there should only ever be a single aggregated `Attestation` for any given
+ * `AttestationData`.
+ *
+ * The pool has a capacity for `SLOTS_RETAINED` slots, when a new `attestation.data.slot` is
+ * provided, the oldest slot is dropped and replaced with the new slot. The pool can also be
+ * pruned by supplying a `current_slot`; all existing attestations with a slot lower than
+ * `current_slot - SLOTS_RETAINED` will be removed and any future attestation with a slot lower
+ * than that will also be refused. Pruning is done automatically based upon the attestations it
+ * receives and it can be triggered manually.
+ */
+export class AttestationPool {
+  private readonly aggregateByRootBySlot = new Map<phase0.Slot, Map<DataRootHex, AggregateFast>>();
+  private lowestPermissibleSlot = 0;
+
+  // TODO: Add metrics for total num of attestations in the pool
+
+  /**
+   * Accepts an `VerifiedUnaggregatedAttestation` and attempts to apply it to the "naive
+   * aggregation pool".
+   *
+   * The naive aggregation pool is used by local validators to produce
+   * `SignedAggregateAndProof`.
+   *
+   * If the attestation is too old (low slot) to be included in the pool it is simply dropped
+   * and no error is returned.
+   *
+   * Expects the attestation to be fully validated:
+   * - Valid signature
+   * - Consistent bitlength
+   * - Valid committeeIndex
+   * - Valid data
+   */
+  add(attestation: phase0.Attestation): void {
+    const slot = attestation.data.slot;
+    const lowestPermissibleSlot = this.lowestPermissibleSlot;
+
+    // Reject any attestations that are too old.
+    if (slot < lowestPermissibleSlot) {
+      throw new AttestationPoolError({code: AttestationPoolErrorCode.SLOT_TOO_LOW, slot, lowestPermissibleSlot});
+    }
+
+    const dataRoot = ssz.phase0.AttestationData.hashTreeRoot(attestation.data);
```

Comment:
wouldn't more future safe way be getForkTypes(slot).AttestationData?

Reply:
Yes but we should change it every at once in another PR. This logic in unchanged in this PR just moved around.